### PR TITLE
Add Chart Status Check and integrate in Shiny UI (Fix #562)

### LINKED ISSUE #562: Support settings requirements for charts
# Summary

Version 1 of the app documented chart-level settings requirements in [this file](https://github.com/SafetyGraphics/safetyGraphics/blob/master/data-raw/settingsMetadataCharts.csv) (e.g. `outlier-explorer` required `id_col`, `measure_col` and `value_col`). This allowed us to enable or disable a chart based on whether all of the required mappings were present. We should consider whether reimplementing this functionality is worthwhile in v2 (or as a later release). Currently, v2 just tries to draw the chart no matter what. If needed settings are missing, the user generally just sees a blank screen for the chart. 

# Details

I'm thinking we'd capture the chart requirements by expanding the `domain` section in the YAML. For example, the YAML for outlier explorer would look like: 

```
env: safetyGraphics
label: Outlier Explorer - widget
type: htmlwidget
domain: 
  - labs
    - id_col
    - measure_col
    - value_col
    - ... 
package: safetyCharts
export: true
...
```

Then we'd need to add some basic validation to the charting modules. Basically we'd check to see whether all of the required settings are present before deciding to render the chart. We could also show ✅ or ❌  in the charts dropdown to indicate whether the chart is available (like we did in v1). 

I definitely think this would be nice to have, but am on the fence about whether it's needed for v2.0. Curious about what others think - @xni7 @elimillera @samussiah @JimBuchanan @mli1  (and everyone) any thoughts?

### PR DESCRIPTION
**Pull request by jwildfire**
# Overview

- `checkChartStatus()` is in good shape and has some simple unit tests. 
- Updated Shiny App to show check/x for status in chart menu and in the header of the chart page. 

# Test Code 

See `tests/checkChartStatus.R` for functional testing. See code below for updates in UI. 

```
a<-makeChartConfig()
a$aeExplorer$dataSpec<-list(
    aes=c("id_col","bodsys_col"),
    dm=c("id_col","treatment_col")
)

a$hepExplorer$dataSpec<-list(
    labs=c("id_col","measure_col")
)

a$aeTimelines$dataSpec <- list(
    aes = c("notacol")
)

safetyGraphicsApp(charts=a)
```

### CODE REVIEW

**Comment by jwildfire** [comment]
cc: @samussiah @bzkrouse

File: R/getChartStatus.R [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        return(colStatus)    
+    })%>% set_names(names(chart$dataSpec))
+
+    # check to see whether all columns in a domain were valid
+    domainStatus <- colStatus %>%
+        map(~all(unlist(.x))) %>%
+        set_names(names(colStatus))
+
+    # check to see whether all columns in all domains were valid
+    status <- ifelse(all(unlist(domainStatus)),"valid","invalid")
```
Comment (bzkrouse) [inline R/getChartStatus.R:66]:
```suggestion
    status <- ifelse(all(unlist(domainStatus)),TRUE, FALSE)
```

File: R/mod_safetyGraphicsServer.R [hunk 2]
```diff
@@ -48,7 +49,28 @@ safetyGraphicsServer <- function(input, output, session, meta, mapping, domainDa
         )
     )
 
-    #Setting tab
+    # Keep chart status updated
+    charts_status <- reactive({
+        charts %>% purrr::map(function(chart){
+            if(hasName(chart, 'dataSpec')){
```
Comment (bzkrouse) [inline R/mod_safetyGraphicsServer.R:55]: @jwildfire Will this piece be updated to check status on charts w/o the dataSpec input as they are configured manually on mapping tab?

Comment (jwildfire) [reply R/mod_safetyGraphicsServer.R:55]: My plan was just to ignore chart objects where `dataSpec` isn't defined. The behavior for those charts would be the same as in v2.1. Basically, they'd try to re-render whenever mapping updates, and if they bomb they bomb ... 

I'm open to suggestions though!

**Comment by jwildfire** [comment]
Going to try to avoid using shinyjs in tab titles via [this approach](https://stackoverflow.com/questions/47896844/shiny-dynamically-change-tab-names)

File: R/mod_safetyGraphicsServer.R (hunk 2 above)
Comment (bzkrouse) [reply R/mod_safetyGraphicsServer.R]: @jwildfire Makes sense! I'm sure you explained that to me, but as a user I was expecting them to show. Could it be a future enhancement?

**Comment by jwildfire** [comment]
@bzkrouse @samussiah Ended up doing a fairly significant refactor of the app to simplify the status workflow. Still planning to add a bit more detail and clean up checks etc.

File: R/mod_safetyGraphicsServer.R (hunk 2 above)
Comment (jwildfire) [reply R/mod_safetyGraphicsServer.R]: I think my problem is that I don't have any good ideas about how to check status if no `dataSpec` is included in the chart object ... We could have an icon or note in the UI indicating that no checking is done for some charts, but not sure what to do beyond that.

**Comment by jwildfire** [comment]
Think I'm pretty much done with this. Kept the UI simple for now (added a status icon with some details on mouseover) - can consider making it more complex in a future release (Modal? Separate collapsable Shiny Module?).

**Comment by jwildfire** [comment]
@samussiah Just bumping this to put it back on your radar.

**Review by samussiah — APPROVED** [review]
This works really nicely @jwildfire! It's nice and visually obvious if anything's off in the mapping. I'll take a look at the corresponding [safetyCharts issue](https://github.com/SafetyGraphics/safetyCharts/issues/111).

**Pull request merged**